### PR TITLE
chore: Introduce the ArtifactMutation type

### PR DESCRIPTION
**Pull request by rumenov**
Have a single interface type for the mutations on replicated artifacts.

### CODE REVIEW

**Review by DSharifi — APPROVED** [review]

**Review by eichhorl — APPROVED** [review]

File: rs/p2p/artifact_manager/src/lib.rs [hunk 1]
```diff
@@ -224,18 +224,25 @@ fn process_messages<Artifact: IdentifiableArtifact + 'static>(
...
+                ArtifactMutation::Insert(artifact_with_opt) => {
+                    let _ = send_advert.blocking_send(ArtifactMutation::Insert(artifact_with_opt));
+                }
+                ArtifactMutation::Remove(id) => purged.push(id),
+            };
         }
+        for id in purged {
+            let _ = send_advert.blocking_send(ArtifactMutation::Remove(id));
+        }
+
```
Comment (tthebst) [inline rs/p2p/artifact_manager/src/lib.rs:245]: You could argue that with this API (`Vec<Mutation>`) it is the job of consensus to order the artifacts in a way that makes sense. The reordering here is a little surprising now whereas before it made sense to send inserts first.

Comment (rumenov) [reply rs/p2p/artifact_manager/src/lib.rs]: I agree it does make more sense. I made the change.

@eichhorl can you PTAL at the MR and confirm it is fine with you as well.

File: rs/artifact_pool/src/ingress_pool.rs [hunk 2]
```diff
@@ -375,9 +375,22 @@ impl MutablePool<SignedIngress> for IngressPoolImpl {
...
+                .drain(..)
+                .filter(|x| {
+                    !purged.contains(&IngressMessageId::new(
+                        x.artifact.expiry_time(),
+                        x.artifact.id(),
+                    ))
+                })
+                .map(ArtifactMutation::Insert),
+        );
+        mutations.extend(purged.drain().map(ArtifactMutation::Remove));
```
Comment (eichhorl) [inline rs/artifact_pool/src/ingress_pool.rs:391]: Given that we now have to do this for every pool, are we sure that what we had before this PR wasn't the better abstraction?

Comment (rumenov) [reply rs/artifact_pool/src/ingress_pool.rs:391]: I personally like the current abstraction (in this PR) better than what we had before. 

Just thinking about how mutations are chained and written on a transaction in DB APIs - the order matters. So in a way we do it in a more idiomatic way with this MR.

E.g. spanner mutations - https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.CommitRequest

Comment (rumenov) [reply rs/artifact_pool/src/ingress_pool.rs:391]: @eichhorl does it make sense?

Comment (eichhorl) [reply rs/artifact_pool/src/ingress_pool.rs:391]: > Just thinking about how mutations are chained and written on a transaction in DB APIs - the order matters.

That makes sense, and I guess the same applies to our pools. Should we return the `ArtifactMutation`s in the same order that we applied the corresponding `ChangeAction`s to the pools?

Comment (rumenov) [reply rs/artifact_pool/src/ingress_pool.rs]: totally!

**Review by eichhorl — APPROVED** [review]

**Pull request merged**